--- a/brave/vendor/bat-native-ledger/src/bat/ledger/internal/contribution/contribution.cc
+++ b/brave/vendor/bat-native-ledger/src/bat/ledger/internal/contribution/contribution.cc
@@ -10,6 +10,7 @@
 #include <memory>
 #include <utility>
 #include <vector>
+#include <ctime>
 
 #include "base/time/time.h"
 #include "bat/ledger/global_constants.h"
--- a/brave/vendor/adblock_rust_ffi/src/wrapper.cpp
+++ b/brave/vendor/adblock_rust_ffi/src/wrapper.cpp
@@ -1,5 +1,6 @@
 #include "wrapper.hpp"
 #include <iostream>
+#include <cstring>
 using namespace std;
 
 extern "C" {
--- a/brave/vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher.cc
+++ b/brave/vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher.cc
@@ -179,7 +179,7 @@
 std::string Publisher::GetBalanceReportName(
     const ledger::ActivityMonth month,
     int year) {
-  return base::StringPrintf("%d_%d", year, month);
+  return base::StringPrintf("%d_%d", year, static_cast<int>(month));
 }

 void Publisher::OnSaveVisitServerPublisher(
