--- b/Telegram/SourceFiles/platform/linux/integration_linux.cpp
+++ a/Telegram/SourceFiles/platform/linux/integration_linux.cpp
@@ -9,6 +9,7 @@
 
 #include "platform/platform_integration.h"
 #include "base/platform/base_platform_info.h"
+#include "base/platform/linux/base_linux_glibmm_helper.h"
 #include "base/platform/linux/base_linux_xdp_utilities.h"
 #include "core/sandbox.h"
 #include "core/application.h"
@@ -124,7 +125,7 @@
 	if (group == "org.freedesktop.appearance"
 		&& key == "color-scheme") {
 		try {
+			const auto ivalue = base::Platform::GlibVariantCast<uint>(value);
-			const auto ivalue = value.get_dynamic<uint>();
 
 			crl::on_main([=] {
 				Core::App().settings().setSystemDarkMode(ivalue == 1);
@@ -265,9 +266,8 @@
 
 	const auto notificationIdVariantType = [] {
 		try {
+			return base::Platform::MakeGlibVariant(
+				NotificationId().toTuple()).get_type();
-			return Glib::create_variant(
-				NotificationId().toTuple()
-			).get_type();
 		} catch (...) {
 			return Glib::VariantType();
 		}
@@ -282,9 +282,9 @@
 					const auto &app = Core::App();
 					app.notifications().manager().notificationActivated(
 						NotificationId::FromTuple(
+							base::Platform::GlibVariantCast<
+								NotificationIdTuple
+							>(parameter)));
-							parameter.get_dynamic<NotificationIdTuple>()
-						)
-					);
 				} catch (...) {
 				}
 			});
@@ -299,10 +299,10 @@
 					const auto &app = Core::App();
 					app.notifications().manager().notificationReplied(
 						NotificationId::FromTuple(
+							base::Platform::GlibVariantCast<
+								NotificationIdTuple
+							>(parameter)),
+						{});
-							parameter.get_dynamic<NotificationIdTuple>()
-						),
-						{}
-					);
 				} catch (...) {
 				}
 			});
--- b/Telegram/SourceFiles/platform/linux/linux_xdp_open_with_dialog.cpp
+++ a/Telegram/SourceFiles/platform/linux/linux_xdp_open_with_dialog.cpp
@@ -8,6 +8,7 @@
 #include "platform/linux/linux_xdp_open_with_dialog.h"
 
 #include "base/platform/base_platform_info.h"
+#include "base/platform/linux/base_linux_glibmm_helper.h"
 #include "base/platform/linux/base_linux_xdp_utilities.h"
 #include "base/platform/linux/base_linux_wayland_integration.h"
 #include "core/application.h"
@@ -33,17 +34,20 @@
 		const auto connection = Gio::DBus::Connection::get_sync(
 			Gio::DBus::BusType::SESSION);
 
+		auto reply = connection->call_sync(
-		const auto version = connection->call_sync(
 			std::string(base::Platform::XDP::kObjectPath),
 			std::string(kPropertiesInterface),
 			"Get",
+			base::Platform::MakeGlibVariant(std::tuple{
-			Glib::create_variant(std::tuple{
 				Glib::ustring(
 					std::string(kXDPOpenURIInterface)),
 				Glib::ustring("version"),
 			}),
+			std::string(base::Platform::XDP::kService));
+
+		const auto version = base::Platform::GlibVariantCast<uint>(
+			base::Platform::GlibVariantCast<Glib::VariantBase>(
+				reply.get_child(0)));
-			std::string(base::Platform::XDP::kService)
-		).get_child(0).get_dynamic<Glib::Variant<uint>>().get();
 
 		if (version < 3) {
 			return false;
@@ -108,7 +112,7 @@
 				loop->quit();
 			},
 			std::string(base::Platform::XDP::kService),
+			"org.freedesktop.portal.Request",
-			std::string(base::Platform::XDP::kRequestInterface),
 			"Response",
 			requestPath);
 
@@ -124,23 +128,26 @@
 			std::string(base::Platform::XDP::kObjectPath),
 			std::string(kXDPOpenURIInterface),
 			"OpenFile",
+			Glib::VariantContainerBase::create_tuple({
+				Glib::Variant<Glib::ustring>::create(parentWindowId),
+				Glib::Variant<int>::create_handle(0),
+				Glib::Variant<std::map<
+					Glib::ustring,
+					Glib::VariantBase
+				>>::create({
-			Glib::create_variant(std::tuple{
-				parentWindowId,
-				Glib::DBusHandle(),
-				std::map<Glib::ustring, Glib::VariantBase>{
 					{
 						"handle_token",
+						Glib::Variant<Glib::ustring>::create(handleToken)
-						Glib::create_variant(handleToken)
 					},
 					{
 						"activation_token",
+						Glib::Variant<Glib::ustring>::create(activationToken)
-						Glib::create_variant(activationToken)
 					},
 					{
 						"ask",
+						Glib::Variant<bool>::create(true)
-						Glib::create_variant(true)
 					},
+				}),
-				},
 			}),
 			Gio::UnixFDList::create(std::vector<int>{ fd }),
 			outFdList,
--- b/Telegram/SourceFiles/platform/linux/main_window_linux.cpp
+++ a/Telegram/SourceFiles/platform/linux/main_window_linux.cpp
@@ -26,6 +26,7 @@
 #include "window/window_controller.h"
 #include "window/window_session_controller.h"
 #include "base/platform/base_platform_info.h"
+#include "base/platform/linux/base_linux_glibmm_helper.h"
 #include "base/event_filter.h"
 #include "ui/widgets/popup_menu.h"
 #include "ui/widgets/input_fields.h"
@@ -250,11 +251,13 @@
 		// According to the spec, it should be of 'x' D-Bus signature,
 		// which corresponds to signed 64-bit integer
 		// https://wiki.ubuntu.com/Unity/LauncherAPI#Low_level_DBus_API:_com.canonical.Unity.LauncherEntry
+		dbusUnityProperties["count"] = Glib::Variant<int64>::create(
+			counterSlice);
+		dbusUnityProperties["count-visible"] =
+			Glib::Variant<bool>::create(true);
-		dbusUnityProperties["count"] = Glib::create_variant(
-			int64(counterSlice));
-		dbusUnityProperties["count-visible"] = Glib::create_variant(true);
 	} else {
+		dbusUnityProperties["count-visible"] =
+			Glib::Variant<bool>::create(false);
-		dbusUnityProperties["count-visible"] = Glib::create_variant(false);
 	}
 
 	try {
@@ -267,7 +270,7 @@
 			"com.canonical.Unity.LauncherEntry",
 			"Update",
 			{},
+			base::Platform::MakeGlibVariant(std::tuple{
-			Glib::create_variant(std::tuple{
 				launcherUrl,
 				dbusUnityProperties,
 			}));
--- b/Telegram/SourceFiles/platform/linux/notifications_manager_linux.cpp
+++ a/Telegram/SourceFiles/platform/linux/notifications_manager_linux.cpp
@@ -10,6 +10,7 @@
 
 #include "base/options.h"
 #include "base/platform/base_platform_info.h"
+#include "base/platform/linux/base_linux_glibmm_helper.h"
 #include "base/platform/linux/base_linux_dbus_utilities.h"
 #include "core/application.h"
 #include "core/sandbox.h"
@@ -42,6 +43,8 @@
 constexpr auto kInterface = kService;
 constexpr auto kPropertiesInterface = "org.freedesktop.DBus.Properties"_cs;
 
+using namespace base::Platform;
+
 struct ServerInformation {
 	QString name;
 	QString vendor;
@@ -113,10 +116,10 @@
 		const auto connection = Gio::DBus::Connection::get_sync(
 			Gio::DBus::BusType::SESSION);
 
+		DBus::StartServiceByNameAsync(
-		base::Platform::DBus::StartServiceByNameAsync(
 			connection,
 			std::string(kService),
+			[=](Fn<DBus::StartReply()> result) {
-			[=](Fn<base::Platform::DBus::StartReply()> result) {
 				Noexcept([&] {
 					try {
 						result(); // get the error if any
@@ -153,7 +156,7 @@
 
 		const auto hasOwner = [&] {
 			try {
+				return DBus::NameHasOwner(
-				return base::Platform::DBus::NameHasOwner(
 					connection,
 					std::string(kService));
 			} catch (...) {
@@ -164,7 +167,7 @@
 		static const auto activatable = [&] {
 			try {
 				return ranges::contains(
+					DBus::ListActivatableNames(connection),
-					base::Platform::DBus::ListActivatableNames(connection),
 					std::string(kService),
 					&Glib::ustring::raw);
 			} catch (...) {
@@ -192,23 +195,19 @@
 			{},
 			[=](const Glib::RefPtr<Gio::AsyncResult> &result) {
 				Noexcept([&] {
+					auto reply = connection->call_finish(result);
-					const auto reply = connection->call_finish(result);
 
+					const auto name = GlibVariantCast<Glib::ustring>(
+						reply.get_child(0));
-					const auto name = reply.get_child(
-						0
-					).get_dynamic<Glib::ustring>();
 
+					const auto vendor = GlibVariantCast<Glib::ustring>(
+						reply.get_child(1));
-					const auto vendor = reply.get_child(
-						1
-					).get_dynamic<Glib::ustring>();
 
+					const auto version = GlibVariantCast<Glib::ustring>(
+						reply.get_child(2));
-					const auto version = reply.get_child(
-						2
-					).get_dynamic<Glib::ustring>();
 
+					const auto specVersion = GlibVariantCast<Glib::ustring>(
+						reply.get_child(3));
-					const auto specVersion = reply.get_child(
-						3
-					).get_dynamic<Glib::ustring>();
 
 					crl::on_main([=] {
 						callback(ServerInformation{
@@ -242,17 +241,18 @@
 			{},
 			[=](const Glib::RefPtr<Gio::AsyncResult> &result) {
 				Noexcept([&] {
+					auto reply = connection->call_finish(result);
+
 					QStringList value;
 					ranges::transform(
+						GlibVariantCast<std::vector<Glib::ustring>>(
+							reply.get_child(0)),
-						connection->call_finish(
-							result
-						).get_child(
-							0
-						).get_dynamic<std::vector<Glib::ustring>>(),
 						ranges::back_inserter(value),
 						QString::fromStdString);
 
+					crl::on_main([=] {
+						callback(value);
+					});
-					crl::on_main([=] { callback(value); });
 				}, [&] {
 					crl::on_main([=] { callback({}); });
 				});
@@ -272,20 +272,21 @@
 			std::string(kObjectPath),
 			std::string(kPropertiesInterface),
 			"Get",
+			MakeGlibVariant(std::tuple{
-			Glib::create_variant(std::tuple{
 				Glib::ustring(std::string(kInterface)),
 				Glib::ustring("Inhibited"),
 			}),
 			[=](const Glib::RefPtr<Gio::AsyncResult> &result) {
 				Noexcept([&] {
+					auto reply = connection->call_finish(result);
+
+					const auto value = GlibVariantCast<bool>(
+						GlibVariantCast<Glib::VariantBase>(
+							reply.get_child(0)));
+
+					crl::on_main([=] {
+						callback(value);
+					});
-					const auto value = connection->call_finish(
-						result
-					).get_child(
-						0
-					).get_dynamic<Glib::Variant<bool>>(
-					).get();
-
-					crl::on_main([=] { callback(value); });
 				}, [&] {
 					crl::on_main([=] { callback(false); });
 				});
@@ -469,38 +470,38 @@
 			const Glib::ustring &object_path,
 			const Glib::ustring &interface_name,
 			const Glib::ustring &signal_name,
+			Glib::VariantContainerBase parameters) {
-			const Glib::VariantContainerBase &parameters) {
 		Noexcept([&] {
 			if (signal_name == "ActionInvoked") {
+				const auto id = GlibVariantCast<uint>(
+					parameters.get_child(0));
-				const auto id = parameters.get_child(0).get_dynamic<uint>();
 
+				const auto actionName = GlibVariantCast<Glib::ustring>(
+					parameters.get_child(1));
-				const auto actionName = parameters.get_child(
-					1
-				).get_dynamic<Glib::ustring>();
 
 				crl::on_main(weak, [=] { actionInvoked(id, actionName); });
 			} else if (signal_name == "ActivationToken") {
+				const auto id = GlibVariantCast<uint>(
+					parameters.get_child(0));
-				const auto id = parameters.get_child(0).get_dynamic<uint>();
 
+				const auto token = GlibVariantCast<Glib::ustring>(
+					parameters.get_child(1));
-				const auto token = parameters.get_child(
-					1
-				).get_dynamic<Glib::ustring>();
 
 				crl::on_main(weak, [=] { activationToken(id, token); });
 			} else if (signal_name == "NotificationReplied") {
+				const auto id = GlibVariantCast<uint>(
+					parameters.get_child(0));
-				const auto id = parameters.get_child(0).get_dynamic<uint>();
 
+				const auto text = GlibVariantCast<Glib::ustring>(
+					parameters.get_child(1));
-				const auto text = parameters.get_child(
-					1
-				).get_dynamic<Glib::ustring>();
 
 				crl::on_main(weak, [=] { notificationReplied(id, text); });
 			} else if (signal_name == "NotificationClosed") {
+				const auto id = GlibVariantCast<uint>(
+					parameters.get_child(0));
-				const auto id = parameters.get_child(0).get_dynamic<uint>();
 
+				const auto reason = GlibVariantCast<uint>(
+					parameters.get_child(1));
-				const auto reason = parameters.get_child(
-					1
-				).get_dynamic<uint>();
 
 				crl::on_main(weak, [=] { notificationClosed(id, reason); });
 			}
@@ -567,30 +568,30 @@
 	}
 
 	if (capabilities.contains("action-icons")) {
+		_hints["action-icons"] = Glib::Variant<bool>::create(true);
-		_hints["action-icons"] = Glib::create_variant(true);
 	}
 
 	// suppress system sound if telegram sound activated,
 	// otherwise use system sound
 	if (capabilities.contains("sound")) {
 		if (Core::App().settings().soundNotify()) {
+			_hints["suppress-sound"] = Glib::Variant<bool>::create(true);
-			_hints["suppress-sound"] = Glib::create_variant(true);
 		} else {
 			// sound name according to http://0pointer.de/public/sound-naming-spec.html
+			_hints["sound-name"] = Glib::Variant<Glib::ustring>::create(
+				"message-new-instant");
-			_hints["sound-name"] = Glib::create_variant(
-				Glib::ustring("message-new-instant"));
 		}
 	}
 
 	if (capabilities.contains("x-canonical-append")) {
+		_hints["x-canonical-append"] = Glib::Variant<Glib::ustring>::create(
+			"true");
-		_hints["x-canonical-append"] = Glib::create_variant(
-			Glib::ustring("true"));
 	}
 
+	_hints["category"] = Glib::Variant<Glib::ustring>::create("im.received");
-	_hints["category"] = Glib::create_variant(Glib::ustring("im.received"));
 
+	_hints["desktop-entry"] = Glib::Variant<Glib::ustring>::create(
+		QGuiApplication::desktopFileName().toStdString());
-	_hints["desktop-entry"] = Glib::create_variant(
-		Glib::ustring(QGuiApplication::desktopFileName().toStdString()));
 
 	_notificationClosedSignalId = _dbusConnection->signal_subscribe(
 		signalEmitted,
@@ -646,7 +647,7 @@
 			std::string(kObjectPath),
 			std::string(kInterface),
 			"Notify",
+			MakeGlibVariant(std::tuple{
-			Glib::create_variant(std::tuple{
 				Glib::ustring(std::string(AppName)),
 				uint(0),
 				iconName,
@@ -658,12 +659,9 @@
 			}),
 			[=](const Glib::RefPtr<Gio::AsyncResult> &result) {
 				Noexcept([&] {
+					auto reply = connection->call_finish(result);
+					const auto notificationId = GlibVariantCast<uint>(
+						reply.get_child(0));
-					const auto notificationId = connection->call_finish(
-						result
-					).get_child(
-						0
-					).get_dynamic<uint>();
-
 					crl::on_main(weak, [=] {
 						_notificationId = notificationId;
 					});
@@ -693,7 +691,7 @@
 		std::string(kObjectPath),
 		std::string(kInterface),
 		"CloseNotification",
+		MakeGlibVariant(std::tuple{
-		Glib::create_variant(std::tuple{
 			_notificationId,
 		}),
 		{},
@@ -730,7 +728,7 @@
 		? image.convertToFormat(QImage::Format_RGBA8888)
 		: image.convertToFormat(QImage::Format_RGB888);
 
+	_hints[_imageKey] = MakeGlibVariant(std::tuple{
-	_hints[_imageKey] = Glib::create_variant(std::tuple{
 		convertedImage.width(),
 		convertedImage.height(),
 		int(convertedImage.bytesPerLine()),
@@ -981,22 +979,19 @@
 					const Glib::ustring &object_path,
 					const Glib::ustring &interface_name,
 					const Glib::ustring &signal_name,
+					Glib::VariantContainerBase parameters) {
-					const Glib::VariantContainerBase &parameters) {
 				Noexcept([&] {
+					const auto interface = GlibVariantCast<Glib::ustring>(
+						parameters.get_child(0));
-					const auto interface = parameters.get_child(
-						0
-					).get_dynamic<Glib::ustring>();
 
 					if (interface != kInterface.data()) {
 						return;
 					}
 
+					const auto inhibited = GlibVariantCast<bool>(
+						GlibVariantCast<
+							std::map<Glib::ustring, Glib::VariantBase>
+					>(parameters.get_child(1)).at("Inhibited"));
-					const auto inhibited = parameters.get_child(
-						1
-					).get_dynamic<std::map<Glib::ustring, Glib::VariantBase>>(
-					).at(
-						"Inhibited"
-					).get_dynamic<bool>();
 
 					crl::on_main(weak, [=] {
 						_inhibited = inhibited;
--- b/Telegram/SourceFiles/platform/linux/specific_linux.cpp
+++ a/Telegram/SourceFiles/platform/linux/specific_linux.cpp
@@ -9,6 +9,7 @@
 
 #include "base/random.h"
 #include "base/platform/base_platform_info.h"
+#include "base/platform/linux/base_linux_glibmm_helper.h"
 #include "base/platform/linux/base_linux_dbus_utilities.h"
 #include "base/platform/linux/base_linux_xdp_utilities.h"
 #include "platform/linux/linux_desktop_environment.h"
@@ -89,13 +90,13 @@
 		commandline.push_back("-autostart");
 
 		std::map<Glib::ustring, Glib::VariantBase> options;
+		options["handle_token"] = Glib::Variant<Glib::ustring>::create(
+			handleToken);
+		options["reason"] = Glib::Variant<Glib::ustring>::create(
+			tr::lng_settings_auto_start(tr::now).toStdString());
+		options["autostart"] = Glib::Variant<bool>::create(start);
+		options["commandline"] = base::Platform::MakeGlibVariant(commandline);
+		options["dbus-activatable"] = Glib::Variant<bool>::create(false);
-		options["handle_token"] = Glib::create_variant(handleToken);
-		options["reason"] = Glib::create_variant(
-			Glib::ustring(
-				tr::lng_settings_auto_start(tr::now).toStdString()));
-		options["autostart"] = Glib::create_variant(start);
-		options["commandline"] = Glib::create_variant(commandline);
-		options["dbus-activatable"] = Glib::create_variant(false);
 
 		auto uniqueName = connection->get_unique_name();
 		uniqueName.erase(0, 1);
@@ -116,11 +117,10 @@
 				const Glib::ustring &object_path,
 				const Glib::ustring &interface_name,
 				const Glib::ustring &signal_name,
+				Glib::VariantContainerBase parameters) {
-				const Glib::VariantContainerBase &parameters) {
 				try {
+					const auto response = base::Platform::GlibVariantCast<
+						uint>(parameters.get_child(0));
-					const auto response = parameters.get_child(
-						0
-					).get_dynamic<uint>();
 
 					if (response) {
 						if (!silent) {
@@ -139,7 +139,7 @@
 				loop->quit();
 			},
 			std::string(base::Platform::XDP::kService),
+			"org.freedesktop.portal.Request",
-			std::string(base::Platform::XDP::kRequestInterface),
 			"Response",
 			requestPath);
 
@@ -153,7 +153,7 @@
 			std::string(base::Platform::XDP::kObjectPath),
 			"org.freedesktop.portal.Background",
 			"RequestBackground",
+			base::Platform::MakeGlibVariant(std::tuple{
-			Glib::create_variant(std::tuple{
 				parentWindowId,
 				options,
 			}),
@@ -429,7 +429,8 @@
 			"color-scheme");
 
 		if (result.has_value()) {
+			const auto value = base::Platform::GlibVariantCast<uint>(*result);
+			return value == 1;
-			return result->get_dynamic<uint>() == 1;
 		}
 	} catch (...) {
 	}
--- b/cmake/external/glibmm/CMakeLists.txt
+++ a/cmake/external/glibmm/CMakeLists.txt
@@ -8,8 +8,8 @@
 add_library(desktop-app::external_glibmm ALIAS external_glibmm)
 
 find_package(PkgConfig REQUIRED REQUIRED)
+pkg_check_modules(GLIBMM REQUIRED IMPORTED_TARGET REQUIRED glibmm-2.68>=2.76)
+pkg_check_modules(GIOMM REQUIRED IMPORTED_TARGET REQUIRED giomm-2.68>=2.76)
-pkg_check_modules(GLIBMM REQUIRED IMPORTED_TARGET REQUIRED glibmm-2.68>=2.77)
-pkg_check_modules(GIOMM REQUIRED IMPORTED_TARGET REQUIRED giomm-2.68>=2.77)
 
 if (DESKTOP_APP_USE_PACKAGED)
     target_link_libraries(external_glibmm
--- b/Telegram/lib_base/CMakeLists.txt
+++ a/Telegram/lib_base/CMakeLists.txt
@@ -35,6 +35,7 @@
     base/platform/linux/base_linux_app_launch_context.h
     base/platform/linux/base_linux_dbus_utilities.cpp
     base/platform/linux/base_linux_dbus_utilities.h
+    base/platform/linux/base_linux_glibmm_helper.h
     base/platform/linux/base_linux_library.cpp
     base/platform/linux/base_linux_library.h
     base/platform/linux/base_linux_wayland_integration.cpp
--- b/Telegram/lib_base/base/platform/linux/base_file_utilities_linux.cpp
+++ a/Telegram/lib_base/base/platform/linux/base_file_utilities_linux.cpp
@@ -7,7 +7,6 @@
 #include "base/platform/linux/base_file_utilities_linux.h"
 
 #include "base/platform/base_platform_file_utilities.h"
-#include "base/platform/linux/base_linux_xdp_utilities.h"
 #include "base/platform/linux/base_linux_app_launch_context.h"
 #include "base/platform/linux/base_linux_wayland_integration.h"
 #include "base/algorithm.h"
@@ -58,22 +57,25 @@
 		auto outFdList = Glib::RefPtr<Gio::UnixFDList>();
 
 		connection->call_sync(
+			"/org/freedesktop/portal/desktop",
-			std::string(XDP::kObjectPath),
 			"org.freedesktop.portal.OpenURI",
 			"OpenDirectory",
+			Glib::VariantContainerBase::create_tuple({
+				Glib::Variant<Glib::ustring>::create({}),
+				Glib::Variant<int>::create_handle(0),
+				Glib::Variant<std::map<
+					Glib::ustring,
+					Glib::VariantBase
+				>>::create({
-			Glib::create_variant(std::tuple{
-				Glib::ustring(),
-				Glib::DBusHandle(),
-				std::map<Glib::ustring, Glib::VariantBase>{
 					{
 						"activation_token",
+						Glib::Variant<Glib::ustring>::create(activationToken)
-						Glib::create_variant(activationToken)
 					},
+				}),
-				},
 			}),
 			Gio::UnixFDList::create(std::vector<int>{ fd }),
 			outFdList,
+			"org.freedesktop.portal.Desktop");
-			std::string(XDP::kService));
 
 		return true;
 	} catch (...) {
@@ -98,11 +100,11 @@
 			"/org/freedesktop/FileManager1",
 			"org.freedesktop.FileManager1",
 			"ShowItems",
+			Glib::VariantContainerBase::create_tuple({
+				Glib::Variant<std::vector<Glib::ustring>>::create({
-			Glib::create_variant(std::tuple{
-				std::vector<Glib::ustring>{
 					Glib::filename_to_uri(filepath.toStdString())
+				}),
+				Glib::Variant<Glib::ustring>::create(startupId),
-				},
-				startupId,
 			}),
 			"org.freedesktop.FileManager1");
 
--- b/Telegram/lib_base/base/platform/linux/base_last_input_linux.cpp
+++ a/Telegram/lib_base/base/platform/linux/base_last_input_linux.cpp
@@ -7,6 +7,7 @@
 #include "base/platform/linux/base_last_input_linux.h"
 
 #include "base/platform/base_platform_info.h"
+#include "base/platform/linux/base_linux_glibmm_helper.h"
 #include "base/debug_log.h"
 
 #ifndef DESKTOP_APP_DISABLE_X11_INTEGRATION
@@ -77,14 +78,14 @@
 			return std::nullopt;
 		}
 
+		auto reply = connection->call_sync(
-		const auto value = connection->call_sync(
 			"/org/freedesktop/ScreenSaver",
 			"org.freedesktop.ScreenSaver",
 			"GetSessionIdleTime",
 			{},
+			"org.freedesktop.ScreenSaver");
-			"org.freedesktop.ScreenSaver"
-		).get_child(0).get_dynamic<uint>();
 
+		const auto value = GlibVariantCast<uint>(reply.get_child(0));
 		return (crl::now() - static_cast<crl::time>(value));
 	} catch (const Glib::Error &e) {
 		static const auto NotSupportedErrors = {
@@ -140,14 +141,14 @@
 			return std::nullopt;
 		}
 
+		auto reply = connection->call_sync(
-		const auto value = connection->call_sync(
 			"/org/gnome/Mutter/IdleMonitor/Core",
 			"org.gnome.Mutter.IdleMonitor",
 			"GetIdletime",
 			{},
+			"org.gnome.Mutter.IdleMonitor");
-			"org.gnome.Mutter.IdleMonitor"
-		).get_child(0).get_dynamic<uint64>();
 
+		const auto value = GlibVariantCast<uint64>(reply.get_child(0));
 		return (crl::now() - static_cast<crl::time>(value));
 	} catch (const Glib::Error &e) {
 		static const auto NotSupportedErrors = {
--- b/Telegram/lib_base/base/platform/linux/base_linux_dbus_utilities.cpp
+++ a/Telegram/lib_base/base/platform/linux/base_linux_dbus_utilities.cpp
@@ -6,43 +6,47 @@
 //
 #include "base/platform/linux/base_linux_dbus_utilities.h"
 
+#include "base/platform/linux/base_linux_glibmm_helper.h"
+
 namespace base::Platform::DBus {
 
 bool NameHasOwner(
 		const Glib::RefPtr<Gio::DBus::Connection> &connection,
 		const Glib::ustring &name) {
+	auto reply = connection->call_sync(
-	return connection->call_sync(
 		std::string(kDBusObjectPath),
 		std::string(kDBusInterface),
 		"NameHasOwner",
+		MakeGlibVariant(std::tuple{name}),
+		std::string(kDBusService));
+
+	return GlibVariantCast<bool>(reply.get_child(0));
-		Glib::create_variant(std::tuple{name}),
-		std::string(kDBusService)
-	).get_child(0).get_dynamic<bool>();
 }
 
 std::vector<Glib::ustring> ListActivatableNames(
 		const Glib::RefPtr<Gio::DBus::Connection> &connection) {
+	auto reply = connection->call_sync(
-	return connection->call_sync(
 		std::string(kDBusObjectPath),
 		std::string(kDBusInterface),
 		"ListActivatableNames",
 		{},
+		std::string(kDBusService));
+
+	return GlibVariantCast<std::vector<Glib::ustring>>(
+		reply.get_child(0));
-		std::string(kDBusService)
-	).get_child(0).get_dynamic<std::vector<Glib::ustring>>();
 }
 
 StartReply StartServiceByName(
 		const Glib::RefPtr<Gio::DBus::Connection> &connection,
 		const Glib::ustring &name) {
+	auto reply = connection->call_sync(
+		std::string(kDBusObjectPath),
+		std::string(kDBusInterface),
+		"StartServiceByName",
+		MakeGlibVariant(std::tuple{ name, uint(0) }),
+		std::string(kDBusService));
+
+	return StartReply(GlibVariantCast<uint>(reply.get_child(0)));
-	return StartReply(
-		connection->call_sync(
-			std::string(kDBusObjectPath),
-			std::string(kDBusInterface),
-			"StartServiceByName",
-			Glib::create_variant(std::tuple{ name, uint(0) }),
-			std::string(kDBusService)
-		).get_child(0).get_dynamic<uint>()
-	);
 }
 
 void StartServiceByNameAsync(
@@ -53,16 +57,11 @@
 		std::string(kDBusObjectPath),
 		std::string(kDBusInterface),
 		"StartServiceByName",
+		MakeGlibVariant(std::tuple{ name, uint(0) }),
-		Glib::create_variant(std::tuple{ name, uint(0) }),
 		[=](const Glib::RefPtr<Gio::AsyncResult> &result) {
 			callback([=] {
+				auto reply = connection->call_finish(result);
+				return StartReply(GlibVariantCast<uint>(reply.get_child(0)));
-				return StartReply(
-					connection->call_finish(
-						result
-					).get_child(
-						0
-					).get_dynamic<uint>()
-				);
 			});
 		},
 		std::string(kDBusService));
@@ -82,19 +81,16 @@
 			const Glib::ustring &object_path,
 			const Glib::ustring &interface_name,
 			const Glib::ustring &signal_name,
+			Glib::VariantContainerBase parameters) {
-			const Glib::VariantContainerBase &parameters) {
 			try {
+				const auto name = GlibVariantCast<Glib::ustring>(
+					parameters.get_child(0));
-				const auto name = parameters.get_child(
-					0
-				).get_dynamic<Glib::ustring>();
 
+				const auto oldOwner = GlibVariantCast<Glib::ustring>(
+					parameters.get_child(1));
-				const auto oldOwner =  parameters.get_child(
-					1
-				).get_dynamic<Glib::ustring>();
 
+				const auto newOwner = GlibVariantCast<Glib::ustring>(
+					parameters.get_child(2));
-				const auto newOwner =  parameters.get_child(
-					2
-				).get_dynamic<Glib::ustring>();
 
 				if (name != service) {
 					return;
--- /dev/null
+++ a/Telegram/lib_base/base/platform/linux/base_linux_glibmm_helper.h
@@ -0,0 +1,130 @@
+// This file is part of Desktop App Toolkit,
+// a set of libraries for developing nice desktop applications.
+//
+// For license and copyright information please follow this link:
+// https://github.com/desktop-app/legal/blob/master/LEGAL
+//
+#pragma once
+
+#include "base/basic_types.h"
+
+#include <glibmm/variant.h>
+
+namespace Glib {
+
+struct Int64Helper0 {};
+using Int64Helper = std::conditional_t<std::is_same_v<gint64, int64>, Int64Helper0, int64>;
+
+struct UInt64Helper0 {};
+using UInt64Helper = std::conditional_t<std::is_same_v<guint64, uint64>, UInt64Helper0, uint64>;
+
+template <>
+class Variant<Int64Helper> : public VariantBase {
+public:
+	using CType = gint64;
+
+	/// Default constructor.
+	Variant() : VariantBase() {
+	}
+
+	/** GVariant constructor.
+	* @param castitem The GVariant to wrap.
+	* @param take_a_reference Whether to take an extra reference of the
+	* GVariant or not (not taking one could destroy the GVariant with the
+	* wrapper).
+	*/
+	explicit Variant(GVariant *castitem, bool take_a_reference = false)
+	: VariantBase(castitem, take_a_reference) {
+	}
+
+	/** Gets the Glib::VariantType.
+	* @return The Glib::VariantType.
+	*/
+	static const VariantType &variant_type() G_GNUC_CONST {
+		static VariantType type(G_VARIANT_TYPE_INT64);
+		return type;
+	}
+
+	/** Creates a new Glib::Variant<int64>.
+	* @param data The value of the new Glib::Variant<int64>.
+	* @return The new Glib::Variant<int64>.
+	*/
+	static Variant<int64> create(int64 data) {
+		auto result = Variant<int64>(g_variant_new_int64(data));
+		return result;
+	}
+
+	/** Gets the value of the Glib::Variant<int64>.
+	* @return The int64 value of the Glib::Variant<int64>.
+	*/
+	int64 get() const {
+		return g_variant_get_int64(gobject_);
+	}
+};
+
+template <>
+class Variant<UInt64Helper> : public VariantBase {
+public:
+	using CType = guint64;
+
+	/// Default constructor.
+	Variant() : VariantBase() {
+	}
+
+	/** GVariant constructor.
+	* @param castitem The GVariant to wrap.
+	* @param take_a_reference Whether to take an extra reference of the
+	* GVariant or not (not taking one could destroy the GVariant with the
+	* wrapper).
+	*/
+	explicit Variant(GVariant *castitem, bool take_a_reference = false)
+	: VariantBase(castitem, take_a_reference) {
+	}
+
+	/** Gets the Glib::VariantType.
+	* @return The Glib::VariantType.
+	*/
+	static const VariantType &variant_type() G_GNUC_CONST {
+		static VariantType type(G_VARIANT_TYPE_UINT64);
+		return type;
+	}
+
+	/** Creates a new Glib::Variant<uint64>.
+	* @param data The value of the new Glib::Variant<uint64>.
+	* @return The new Glib::Variant<uint64>.
+	*/
+	static Variant<uint64> create(uint64 data) {
+		auto result = Variant<uint64>(g_variant_new_uint64(data));
+		return result;
+	}
+
+	/** Gets the value of the Glib::Variant<uint64>.
+	* @return The uint64 value of the Glib::Variant<uint64>.
+	*/
+	uint64 get() const {
+		return g_variant_get_uint64(gobject_);
+	}
+};
+
+} // namespace Glib
+
+namespace base {
+namespace Platform {
+
+template <typename T>
+auto MakeGlibVariant(T &&data) {
+	return Glib::Variant<std::decay_t<T>>::create(data);
+}
+
+template <typename T>
+auto MakeGlibVariant(const T &data) {
+	return Glib::Variant<T>::create(data);
+}
+
+template <typename T>
+auto GlibVariantCast(const Glib::VariantBase &data) {
+	return Glib::VariantBase::cast_dynamic<Glib::Variant<T>>(data).get();
+}
+
+} // namespace Platform
+} // namespace base
--- b/Telegram/lib_base/base/platform/linux/base_linux_xdp_utilities.cpp
+++ a/Telegram/lib_base/base/platform/linux/base_linux_xdp_utilities.cpp
@@ -6,6 +6,7 @@
 //
 #include "base/platform/linux/base_linux_xdp_utilities.h"
 
+#include "base/platform/linux/base_linux_glibmm_helper.h"
 #include "base/platform/linux/base_linux_wayland_integration.h"
 #include "base/platform/base_platform_info.h"
 
@@ -38,21 +39,21 @@
 		const Glib::ustring &group,
 		const Glib::ustring &key) {
 	try {
+		const auto connection = Gio::DBus::Connection::get_sync(
+			Gio::DBus::BusType::SESSION);
+
+		auto reply = connection->call_sync(
-		return Gio::DBus::Connection::get_sync(
-			Gio::DBus::BusType::SESSION
-		)->call_sync(
 			std::string(kObjectPath),
 			std::string(kSettingsInterface),
 			"Read",
+			MakeGlibVariant(std::tuple{
-			Glib::create_variant(std::tuple{
 				group,
 				key,
 			}),
+			std::string(kService));
+
+		return GlibVariantCast<Glib::VariantBase>(
+			GlibVariantCast<Glib::VariantBase>(reply.get_child(0)));
-			std::string(kService)
-		).get_child(
-			0
-		).get_dynamic<Glib::Variant<Glib::VariantBase>>(
-		).get();
 	} catch (...) {
 	}
 
@@ -85,19 +86,16 @@
 				const Glib::ustring &object_path,
 				const Glib::ustring &interface_name,
 				const Glib::ustring &signal_name,
+				Glib::VariantContainerBase parameters) {
-				const Glib::VariantContainerBase &parameters) {
 				try {
+					const auto group = GlibVariantCast<Glib::ustring>(
+						parameters.get_child(0));
-					const auto group = parameters.get_child(
-						0
-					).get_dynamic<Glib::ustring>();
 
+					const auto key = GlibVariantCast<Glib::ustring>(
+						parameters.get_child(1));
-					const auto key = parameters.get_child(
-						1
-					).get_dynamic<Glib::ustring>();
 
+					const auto value = GlibVariantCast<Glib::VariantBase>(
+						parameters.get_child(2));
-					const auto value = parameters.get_child(
-						2
-					).get_dynamic<Glib::VariantBase>();
 
 					callback(group, key, value);
 				} catch (...) {
--- b/Telegram/lib_base/base/platform/linux/base_power_save_blocker_linux.cpp
+++ a/Telegram/lib_base/base/platform/linux/base_power_save_blocker_linux.cpp
@@ -89,20 +89,24 @@
 			std::string(XDP::kObjectPath),
 			"org.freedesktop.portal.Inhibit",
 			"Inhibit",
+			Glib::VariantContainerBase::create_tuple({
+				Glib::Variant<Glib::ustring>::create(
+					XDP::ParentWindowID(window)),
+				Glib::Variant<uint>::create(4), // Suspend
+				Glib::Variant<std::map<
+					Glib::ustring,
+					Glib::VariantBase
+				>>::create({
-			Glib::create_variant(std::tuple{
-				XDP::ParentWindowID(window),
-				uint(4), // Suspend
-				std::map<Glib::ustring, Glib::VariantBase>{
 					{
 						"handle_token",
+						Glib::Variant<Glib::ustring>::create(handleToken)
-						Glib::create_variant(handleToken)
 					},
 					{
 						"reason",
+						Glib::Variant<Glib::ustring>::create(
+							description.toStdString())
-						Glib::create_variant(
-							Glib::ustring(description.toStdString()))
 					},
+				}),
-				},
 			}),
 			{},
 			std::string(XDP::kService));
--- b/Telegram/lib_base/base/platform/linux/base_url_scheme_linux.cpp
+++ a/Telegram/lib_base/base/platform/linux/base_url_scheme_linux.cpp
@@ -8,6 +8,7 @@
 
 #include "base/const_string.h"
 #include "base/debug_log.h"
+#include "base/platform/linux/base_linux_glibmm_helper.h"
 
 #include <QtGui/QGuiApplication>
 #include <QtWidgets/QWidget>
@@ -30,16 +31,18 @@
 		const auto connection = Gio::DBus::Connection::get_sync(
 			Gio::DBus::BusType::SESSION);
 
+		auto reply = connection->call_sync(
-		const auto currentHandler = connection->call_sync(
 			std::string(kSnapcraftSettingsObjectPath),
 			std::string(kSnapcraftSettingsInterface),
 			"GetSub",
+			MakeGlibVariant(std::tuple{
-			Glib::create_variant(std::tuple{
 				Glib::ustring("default-url-scheme-handler"),
 				Glib::ustring(protocol.toStdString()),
 			}),
+			std::string(kSnapcraftSettingsService));
+
+		const auto currentHandler = GlibVariantCast<Glib::ustring>(
+			reply.get_child(0));
-			std::string(kSnapcraftSettingsService)
-		).get_child(0).get_dynamic<Glib::ustring>();
 
 		const auto expectedHandler = qEnvironmentVariable("SNAP_NAME")
 			+ ".desktop";
@@ -57,7 +60,7 @@
 			std::string(kSnapcraftSettingsObjectPath),
 			std::string(kSnapcraftSettingsInterface),
 			"SetSub",
+			MakeGlibVariant(std::tuple{
-			Glib::create_variant(std::tuple{
 				Glib::ustring("default-url-scheme-handler"),
 				Glib::ustring(protocol.toStdString()),
 				Glib::ustring(expectedHandler.toStdString()),
--- b/Telegram/lib_ui/ui/platform/linux/ui_window_title_linux.cpp
+++ a/Telegram/lib_ui/ui/platform/linux/ui_window_title_linux.cpp
@@ -6,14 +6,13 @@
 //
 #include "ui/platform/linux/ui_window_title_linux.h"
 
+#include "base/platform/linux/base_linux_glibmm_helper.h"
 #include "base/platform/linux/base_linux_xdp_utilities.h"
 
 #ifndef DESKTOP_APP_DISABLE_X11_INTEGRATION
 #include "base/platform/linux/base_linux_xsettings.h"
 #endif // !DESKTOP_APP_DISABLE_X11_INTEGRATION
 
-#include <glibmm.h>
-
 namespace Ui {
 namespace Platform {
 namespace internal {
@@ -121,7 +120,8 @@
 
 			return GtkKeywordsToTitleControlsLayout(
 				QString::fromStdString(
+					base::Platform::GlibVariantCast<Glib::ustring>(
+						*decorationLayout)));
-					decorationLayout->get_dynamic<Glib::ustring>()));
 		} catch (...) {
 		}
 
